src/codegen/events/substrate.rs

```diff
@@ -32,68 +45,121 @@ impl EventEmitter for SubstrateEventEmitter<'_> {
-                                    vec![Type::DynamicBytes],
-                                    ast::Builtin::AbiEncodePacked,
-                                    vec![arg.clone()],
-                                )],
-                            ),
-                            cfg,
-                            contract_no,
-                            Some(func),
-                            self.ns,
-                            vartab,
-                            opt,
-                        );
-
-                        topics.push(e);
-                        topic_tys.push(Type::Bytes(32));
-                    }
-                    _ => {
-                        let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                        topics.push(e);
-                        topic_tys.push(ty);
-                    }
-                }
-            } else {
-                let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                data.push(e);
-                data_tys.push(arg.ty());
+        let loc = pt::Loc::Builtin;
+        let event = &self.ns.events[self.event_no];
+        let hash_len = || Box::new(Expression::NumberLiteral(loc, Type::Uint(32), 32.into()));
+
+        // Events that are not anonymous always have themselves as a topic.
+        // This is static and can be calculated at compile time.
+        if !event.anonymous {
+            let mut encoded =
+                format!("{}::{}", &self.ns.contracts[contract_no].name, &event.name).encode();
+            encoded[0] = 0; // Set the prefix (there is no prefix for the event topic)
+            topics.push(Expression::AllocDynamicBytes(
+                loc,
+                Type::Slice(Type::Uint(8).into()),
+                hash_len(),
+                Some(topic_hash(&encoded[..])),
+            ));
+            topic_tys.push(Type::DynamicBytes);
+        };
+
+        // Topic prefixes are static and can be calculated at compile time.
+        let mut topic_prefixes: Vec<Vec<u8>> = event
+            .fields
+            .iter()
+            .filter(|field| field.indexed)
+            .map(|field| {
+                format!(
+                    "{}::{}::{}",
+                    &self.ns.contracts[contract_no].name,
+                    &event.name,
+                    &field.name_as_str()
+                )
+                .into_bytes()
+                .encode()
+            })
+            .collect();
+
+        for (ast_exp, field) in self.args.iter().zip(event.fields.iter()) {
+            let value_exp = expression(ast_exp, cfg, contract_no, Some(func), self.ns, vartab, opt);
+            data_tys.push(value_exp.ty());
+            data.push(value_exp);
+
+            if !field.indexed {
+                continue;
             }
+
+            let value_exp = expression(ast_exp, cfg, contract_no, Some(func), self.ns, vartab, opt);
```

Comment:
You are calculating `ast_exp` twice - better to evaluate it once and put it into a variable.